File: modules/core/src/main/java/turbolift/internals/engine/concurrent/atomic/AtomicBoolVH.java
package turbolift.internals.engine.concurrent.atomic;
import java.lang.invoke.VarHandle;
import java.lang.invoke.MethodHandles;


public class AtomicBoolVH {
  private boolean valueVH;
  private static final VarHandle nameVH;

  static {
    try {
      nameVH = MethodHandles.lookup().findVarHandle(AtomicBoolVH.class, "valueVH", boolean.class);
    } catch (ReflectiveOperationException e) {
      throw new ExceptionInInitializerError(e);
    }
  }

  public AtomicBoolVH(boolean a) {
    valueVH = a;
  }

  protected boolean getVH() {
    return (boolean) nameVH.get(this);
  }

  protected void setVH(boolean a) {
    nameVH.set(this, a);
  }

  protected boolean gasVH(boolean a) {
    return (boolean) nameVH.getAndSet(this, a);
  }

  protected boolean casVH(boolean a, boolean b) {
    return nameVH.compareAndSet(this, a, b);
  }

  protected boolean caxVH(boolean a, boolean b) {
    return (boolean) nameVH.compareAndExchange(this, a, b);
  }
}

File: modules/core/src/main/scala/turbolift/internals/engine/concurrent/atomic/AtomicBoolVH.scala
package turbolift.internals.engine.concurrent.atomic
import java.lang.invoke.{VarHandle, MethodHandles}


private[turbolift] class AtomicBoolVH(protected var unsafeValue: Boolean):
  protected def getVH: Boolean = AtomicBoolVH.theVH.get(this)
  
  protected def setVH(a: Boolean): Unit = AtomicBoolVH.theVH.set(this, a)
  
  protected def gasVH(a: Boolean): Boolean = AtomicBoolVH.theVH.getAndSet(this, a)

  protected def casVH(a: Boolean, b: Boolean): Boolean = AtomicBoolVH.theVH.compareAndSet(this, a, b)

  protected def caxVH(a: Boolean, b: Boolean): Boolean = AtomicBoolVH.theVH.compareAndExchange(this, a, b)


private[turbolift] object AtomicBoolVH:
  private val theVH: VarHandle = MethodHandles
    .privateLookupIn(classOf[AtomicBoolVH], MethodHandles.lookup()).nn
    .findVarHandle(classOf[AtomicBoolVH], "unsafeValue", java.lang.Boolean.TYPE).nn
